src/gc.c

```diff
@@ -62,44 +67,98 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
   return ret;
 }
 
-static int GCContext_PeriodicCallback(RedisModuleCtx* ctx, void* privdata) {
-  if (RedisModule_AvoidReplicaTraffic && RedisModule_AvoidReplicaTraffic()) {
-    // If slave trafic is not allow it means that there is a state machine running
-    // we do not want to run any GC w(hich might cause a FORK process to start for example).
-    // Its better to just avoid it.
-    return 1;
+static void stopGC(GCContext* gc) {
+  gc->stopped = 1;
+  if (gc->callbacks.kill) {
+    gc->callbacks.kill(gc->gcCtx);
+  }  
+}
+
+static void timerCallback(RedisModuleCtx* ctx, void* data);
+
+static long long getNextPeriod(GCContext* gc) {
+  struct timespec interval = gc->callbacks.getInterval(gc->gcCtx);
+  long long ms = interval.tv_sec * 1000 + interval.tv_nsec / 1000000;  // convert to millisecond
+  return ms;
+}
+
+static RedisModuleTimerID scheduleNext(GCContext* gc) {
+  if (!RedisModule_CreateTimer) return 0;
+  RedisModuleCtx* ctx = RSDummyContext;
+  long long period = getNextPeriod(gc);
+  return RedisModule_CreateTimer(ctx, period, timerCallback, gc);
+}
+
+static void threadCallback(void* data) {
+  GCContext* gc = data;
+  RedisModuleCtx* ctx = RSDummyContext;
+
+  if (gc->stopped) {
+    return;
   }
 
-  GCContext* gc = privdata;
   int ret = gc->callbacks.periodicCallback(ctx, gc->gcCtx);
 
+  RedisModule_ThreadSafeContextLock(ctx);
   RedisModuleBlockedClient* bClient = BlockClients_pop(&gc->bClients);
   if (bClient) {
     RedisModule_UnblockClient(bClient, NULL);
   }
-  if (gc->timer) {
-    // Timer could've been deleted..
-    RMUtilTimer_SetInterval(gc->timer, gc->callbacks.getInterval(gc->gcCtx));
+  if (!ret || gc->stopped) {
+    stopGC(gc);
+    RedisModule_ThreadSafeContextUnlock(ctx);
+    return;
   }
-  return ret;
+  gc->timerID = scheduleNext(gc);
+  RedisModule_ThreadSafeContextUnlock(ctx);
 }
 
-static void GCContext_OnTerm(void* privdata) {
-  GCContext* gc = privdata;
+static void destroyCallback(void* data) {
+  GCContext* gc = data;
+  RedisModuleCtx* ctx = RSDummyContext;
+  assert(gc->stopped == 1);
+
+  RedisModule_ThreadSafeContextLock(ctx);
+  RedisModule_StopTimer(ctx, gc->timerID, NULL);
```

Comment:
What if you do not have schedule timer here?

Reply:
See below